contracts/insurance/Insurance.sol

```diff
@@ -0,0 +1,119 @@
+pragma experimental "ABIEncoderV2";
+
+import "../interfaces/ITransferDefinition.sol";
+import "../lib/SafeMath.sol";
+import "../lib/LibChannelCrypto.sol";
+
+/// @title InsurancePayment
+/// @notice This contract allows someone to put up a payment as "insurance"
+///         which can be used to guarantee virtually anything. The insurance
+///         can only be unlocked by a specified mediator, and will be refunded
+///         to the guarantor after a specified expiration period, or if
+///         explicitly cancelled by the receiver.
+
+contract Insurance is ITransferDefinition {
+  using LibChannelCrypto for bytes32;
+  using SafeMath for uint256;
+
+  // TransferState contains the payment data
+  struct TransferState {
+    address receiver;
+    address mediator;
+    uint256 collateral;
+    uint256 expiration;
+    bytes32 UUID;
+  }
+
+  // Encapsulates the UUID and payment amount to make signing easier - see below TransferResolver
+  struct ResolverData {
+    uint256 amount;
+    bytes32 UUID;
+  }
+
+  // TransferResolver contains the UUID of the payment, the amount, and a signature
+  struct TransferResolver {
+    ResolverData data;
+    bytes signature;
+  }
+
+  /* solhint-disable */
+  string StateEncoding = "tuple(address receiver, address mediator, uint256 collateral, uint256 expiration, bytes32 UUID)";
+  string ResolverEncoding = "tuple(tuple(uint256 amount, bytes32 UUID) data, bytes signature)";
+  string Name = "Insurance";
+  /* solhint-enable */
+
+  function getRegistryInformation() external override view returns (RegisteredTransfer memory) {
+    RegisteredTransfer memory info = RegisteredTransfer({
+      name: Name,
+      stateEncoding: StateEncoding,
+      resolverEncoding: ResolverEncoding,
+      definition: address(this)
+    });
+    return info;
+  }
+
+  /// @notice Creates an insurancePayment from the guarantor to the receiver
+  /// @param encodedBalance balance the guarantor is putting into the transfer
+  /// @param encodedState encoded transfer parameters (insurance payment info)
+  /// @dev encoded state info contains the receiver & mediator addresses, collateral amount, expiration, and UUID
+  function create(
+		bytes calldata encodedBalance,
+		bytes calldata encodedState
+	) external override view returns (bool) {
+    Balance memory balance = abi.decode(encodedBalance, (Balance));
+    TransferState memory state = abi.decode(encodedState, (TransferState));
+
+    // Receiver should not have an initial balance within the transfer
+    require(balance.amount[1] == 0, "Cannot create parameterized payment with nonzero recipient init balance");
+
+    // Insurance payment checks
+    require(state.expiration > block.timestamp + 3 days, "Expiration must be at least 3 days in the future.");
+
+    // Sanity checks
+    require(state.receiver != address(0x0), "Receiver address cannot be the zero address!");
+    require(state.mediator != address(0x0), "Mediator address cannot be the zero address!");
```

Comment:
Do you also want to ensure the `state.receiver != state.mediator`?

Reply:
Not necessarily. That would be the normal use case, yes, but if someone wants to set the receiver as the mediator, I think we should let them.